Import from remote BibTex libraries cannot handle CR-LF

### Type of JetBrains IDE (IntelliJ, PyCharm, etc.) and version
PyCharm 2024.2.0.1 (build PC-242.20224.428)

### Operating System 
Windows 10.0 (amd64)

### TeXiFy IDEA version
0.9.7

### Description

Apparently,  the automatic import from remote BibTex libraries to local library cannot handle BibTex files that use CR-LF line endings. 
It works fine for just LF line endings.



### Stacktrace
```
java.lang.AssertionError: Wrong line separators: '...Datta2017,\r\n  author...' at offset 19
	at com.intellij.openapi.util.text.StringUtil.assertValidSeparators(StringUtil.java:2552)
	at com.intellij.openapi.editor.impl.DocumentImpl.assertValidSeparators(DocumentImpl.java:716)
	at com.intellij.openapi.editor.impl.DocumentImpl.replaceString(DocumentImpl.java:607)
	at com.intellij.openapi.editor.impl.DocumentImpl.replaceString(DocumentImpl.java:598)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.doCommitTransaction(PsiToDocumentSynchronizer.java:212)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.lambda$commitTransaction$1(PsiToDocumentSynchronizer.java:188)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.lambda$doSync$0(PsiToDocumentSynchronizer.java:106)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.performAtomically(PsiToDocumentSynchronizer.java:124)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.doSync(PsiToDocumentSynchronizer.java:106)
	at com.intellij.psi.impl.PsiToDocumentSynchronizer.commitTransaction(PsiToDocumentSynchronizer.java:188)
	at com.intellij.pom.core.impl.PomModelImpl.commitTransaction(PomModelImpl.java:195)
	at com.intellij.pom.core.impl.PomModelImpl.lambda$runTransaction$1(PomModelImpl.java:151)
	at com.intellij.psi.impl.DebugUtil.performPsiModification(DebugUtil.java:535)
	at com.intellij.pom.core.impl.PomModelImpl.lambda$runTransaction$2(PomModelImpl.java:103)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeNonCancelableSection$3(CoreProgressManager.java:269)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:735)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:691)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$computeInNonCancelableSection$4(CoreProgressManager.java:277)
	at com.intellij.openapi.progress.Cancellation.computeInNonCancelableSection(Cancellation.java:62)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeInNonCancelableSection(CoreProgressManager.java:277)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:268)
	at com.intellij.pom.core.impl.PomModelImpl.runTransaction(PomModelImpl.java:92)
	at com.intellij.psi.impl.source.tree.ChangeUtil.prepareAndRunChangeAction(ChangeUtil.java:141)
	at com.intellij.psi.impl.source.tree.CompositeElement.addChild(CompositeElement.java:586)
	at com.intellij.psi.impl.source.tree.CompositeElement.addChildren(CompositeElement.java:680)
	at com.intellij.psi.impl.source.codeStyle.CodeEditUtil.addChildren(CodeEditUtil.java:53)
	at com.intellij.psi.impl.source.tree.CompositeElement.addInternal(CompositeElement.java:455)
	at com.intellij.psi.impl.source.PsiFileImpl.addAfter(PsiFileImpl.java:887)
	at nl.hannahsten.texifyidea.completion.handlers.LatexReferenceInsertHandler.handleInsert(LatexReferenceInsertHandler.kt:25)
	at com.intellij.codeInsight.lookup.LookupElementBuilder.handleInsert(LookupElementBuilder.java:343)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.lambda$insertItem$15(CodeCompletionHandlerBase.java:713)
	at com.intellij.util.indexing.FileBasedIndex.lambda$ignoreDumbMode$0(FileBasedIndex.java:225)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:111)
	at com.intellij.util.indexing.FileBasedIndexEx.ignoreDumbMode(FileBasedIndexEx.java:645)
	at com.intellij.util.indexing.FileBasedIndex.ignoreDumbMode(FileBasedIndex.java:224)
	at com.intellij.util.indexing.DumbModeAccessType.ignoreDumbMode(DumbModeAccessType.java:41)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.lambda$insertItem$16(CodeCompletionHandlerBase.java:712)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWriteAction$lambda$4(AnyThreadWriteThreadingSupport.kt:318)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWriteAction(AnyThreadWriteThreadingSupport.kt:328)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWriteAction(AnyThreadWriteThreadingSupport.kt:318)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:890)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.insertItem(CodeCompletionHandlerBase.java:696)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.doInsertItemForSingleCaret(CodeCompletionHandlerBase.java:640)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.lambda$doInsertItem$13(CodeCompletionHandlerBase.java:612)
	at com.intellij.openapi.editor.impl.CaretModelImpl.lambda$runForEachCaret$3(CaretModelImpl.java:303)
	at com.intellij.openapi.editor.impl.CaretModelImpl.doWithCaretMerging(CaretModelImpl.java:412)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:312)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:289)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.doInsertItem(CodeCompletionHandlerBase.java:604)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.insertItemHonorBlockSelection(CodeCompletionHandlerBase.java:563)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.lookupItemSelected(CodeCompletionHandlerBase.java:509)
	at com.intellij.codeInsight.completion.CompletionProgressIndicator.itemSelected(CompletionProgressIndicator.java:211)
	at com.intellij.codeInsight.completion.BaseCompletionLookupArranger.itemSelected(BaseCompletionLookupArranger.java:229)
	at com.intellij.codeInsight.lookup.impl.LookupImpl.fireItemSelected(LookupImpl.java:1000)
	at com.intellij.codeInsight.lookup.impl.LookupImpl.finishLookupInWritableFile(LookupImpl.java:617)
	at com.intellij.codeInsight.lookup.impl.LookupImpl.lambda$finishLookup$8(LookupImpl.java:573)
	at 
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 ## [Unreleased]
 
 ### Added
+* Fix auto import from local bibtex file on Windows
 * Add sections to breadcrumbs
 * Change order in structure view to match source file and sectioning level 
 * Add command redefinitions to command definition filter in structure view

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,15 +3,35 @@
 ## [Unreleased]
 
 ### Added
-* Fix auto import from local bibtex file on Windows
+
+### Fixed
+
+## [0.9.10-alpha.4] - 2024-12-21
+
+### Added
+
 * Add sections to breadcrumbs
 * Improve performance when starting a run configuration and when using autocompletion directly after starting the IDE
 * Change order in structure view to match source file and sectioning level 
 * Add command redefinitions to command definition filter in structure view
 * Add support for automatic language injection on the minted environment
 * Add support for automatic detection of custom command aliases for include commands
+* Add support for DeclareGraphicsExtensions
+* Add inspection to warn about a missing reference for a glossary occurrence
+* Do not fold sections in a command definition
+* Include optional parameters in spellcheck, if it contains text
+* Improve performance of finding files to be indexed
+* Show formatted file path in file not found inspection quickfix name
+* Automatically index bibliography files outside the project that are included by an absolute path
+* Disable quotes inspection when TeX ligatures are disabled by fontspec
+* Inspections can now be suppressed for any single line, or block of text
 
 ### Fixed
+
+* Fix auto import from local bibtex file on Windows
+* Fix false positive for duplicate command definition inspection in if/else
+* Fix LaTeX files not showing up when choosing main file in run configuration
+* Fix various issues with the Grazie implementation, in particular default rules for Grazie Pro
 
 ## [0.9.10-alpha.3] - 2024-12-15
 
@@ -28,6 +48,7 @@
 * Inspections can now be suppressed for any single line, or block of text
 
 ### Fixed
+
 * Fix false positive for duplicate command definition inspection in if/else
 * Fix LaTeX files not showing up when choosing main file in run configuration
 * Fix various issues with the Grazie implementation, in particular default rules for Grazie Pro
@@ -479,7 +500,8 @@ Thanks to @jojo2357 and @MisterDeenis for contributing to this release!
 * Fix some intention previews. ([#2796](https://github.com/Hannah-Sten/TeXiFy-IDEA/issues/2796))
 * Other small bug fixes and improvements. ([#2776](https://github.com/Hannah-Sten/TeXiFy-IDEA/issues/2776), [#2774](https://github.com/Hannah-Sten/TeXiFy-IDEA/issues/2774), [#2765](https://github.com/Hannah-Sten/TeXiFy-IDEA/issues/2765)-[#2773](https://github.com/Hannah-Sten/TeXiFy-IDEA/issues/2773))
 
-[Unreleased]: https://github.com/Hannah-Sten/TeXiFy-IDEA/compare/v0.9.10-alpha.3...HEAD
+[Unreleased]: https://github.com/Hannah-Sten/TeXiFy-IDEA/compare/v0.9.10-alpha.4...HEAD
+[0.9.10-alpha.4]: https://github.com/Hannah-Sten/TeXiFy-IDEA/compare/v0.9.10-alpha.3...v0.9.10-alpha.4
 [0.9.10-alpha.3]: https://github.com/Hannah-Sten/TeXiFy-IDEA/compare/v0.9.10-alpha.2...v0.9.10-alpha.3
 [0.9.10-alpha.2]: https://github.com/Hannah-Sten/TeXiFy-IDEA/compare/v0.9.9...v0.9.10-alpha.2
 [0.9.9]: https://github.com/Hannah-Sten/TeXiFy-IDEA/compare/v0.9.8...v0.9.9

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,4 +1,4 @@
-pluginVersion = 0.9.10-alpha.3
+pluginVersion = 0.9.10-alpha.4
 
 #     Info about build ranges: https://www.jetbrains.org/intellij/sdk/docs/basics/getting_started/build_number_ranges.html
 #    Note that an xyz branch corresponds to version 20xy.z and a since build of xyz.*

--- a/src/nl/hannahsten/texifyidea/completion/handlers/LatexReferenceInsertHandler.kt
+++ b/src/nl/hannahsten/texifyidea/completion/handlers/LatexReferenceInsertHandler.kt
@@ -5,9 +5,11 @@ import com.intellij.codeInsight.completion.InsertionContext
 import com.intellij.codeInsight.lookup.LookupElement
 import nl.hannahsten.texifyidea.file.BibtexFile
 import nl.hannahsten.texifyidea.psi.BibtexEntry
+import nl.hannahsten.texifyidea.psi.LatexPsiHelper
 import nl.hannahsten.texifyidea.settings.TexifySettings
 import nl.hannahsten.texifyidea.util.files.bibtexIdsInFileSet
 import nl.hannahsten.texifyidea.util.files.referencedFileSet
+import nl.hannahsten.texifyidea.util.parser.firstChildOfType
 
 /**
  * @author Sten Wessel
@@ -20,10 +22,13 @@ class LatexReferenceInsertHandler(private val remote: Boolean = false, private v
 
         if (remote and TexifySettings.getInstance().automaticBibtexImport) {
             remoteBib ?: return
+            // remoteBib may come from a file with CRLF line separators, which cannot be accepted into a psi file, so we need to fix that
+            val newBibEntry = LatexPsiHelper(context.project).createBibtexFromText(remoteBib.text.replace("\r\n", "\n")).firstChildOfType(BibtexEntry::class) ?: return
+
             val bibsInFile = context.file.originalFile.bibtexIdsInFileSet()
             // Add the bib item after the last item we found in the file set, and hope that that makes sense...
             bibsInFile.lastOrNull()?.let {
-                it.parent.addAfter(remoteBib, it)
+                it.parent.addAfter(newBibEntry, it)
             }
 
             // If there are no bib items in the fileset yet, see if there is a(n empty) bib file we can add the bib entry to.
@@ -31,7 +36,7 @@ class LatexReferenceInsertHandler(private val remote: Boolean = false, private v
                 context.file.originalFile
                     .referencedFileSet()
                     .firstOrNull { it is BibtexFile }
-                    ?.add(remoteBib)
+                    ?.add(newBibEntry)
             }
         }
     }

--- a/src/nl/hannahsten/texifyidea/psi/LatexPsiHelper.kt
+++ b/src/nl/hannahsten/texifyidea/psi/LatexPsiHelper.kt
@@ -6,6 +6,7 @@ import com.intellij.psi.PsiFile
 import com.intellij.psi.PsiFileFactory
 import com.intellij.psi.PsiWhiteSpace
 import com.intellij.psi.impl.source.tree.LeafPsiElement
+import nl.hannahsten.texifyidea.grammar.BibtexLanguage
 import nl.hannahsten.texifyidea.grammar.LatexLanguage
 import nl.hannahsten.texifyidea.psi.LatexTypes.*
 import nl.hannahsten.texifyidea.util.Log
@@ -56,6 +57,9 @@ class LatexPsiHelper(private val project: Project) {
      */
     fun createFromText(text: String): PsiFile =
         PsiFileFactory.getInstance(project).createFileFromText("DUMMY.tex", LatexLanguage, text, false, true)
+
+    fun createBibtexFromText(text: String): PsiFile =
+        PsiFileFactory.getInstance(project).createFileFromText("DUMMY.bib", BibtexLanguage, text, false, true)
 
     /**
      * Adds the supplied element to the content of the environment.

